### DISCUSSION TITLE
Questions on Joint Error in Kinematic Chain

### DISCUSSION THREAD

**Discussion by linhongbin** [surgical-robotics-ai/surgical_robotics_challenge#41] [Q&A] [answered]
Hi,

I want to ask some questions about joint error in the evaluation of Task 2 and 3:

- Is it the value of joint error in the discrete set: {-5., -2., 0 , 2., 5.} in the evaluation? Code implementation can be found [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/a85dd0271d1b11b6f964d5f46908f2f4060ddcdf/scripts/surgical_robotics_challenge/psm_arm.py#L100)

- Are there joint error for only Joint 1 and 2 of PSM in the evaluation? Code implementation can be found [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/a85dd0271d1b11b6f964d5f46908f2f4060ddcdf/scripts/surgical_robotics_challenge/psm_arm.py#L103). I got little bit confused by the [discussion](https://github.com/collaborative-robotics/surgical_robotics_challenge/discussions/36#discussioncomment-2511297) since you mentioned that Joint 3, the insertion joint, will also have the joint error.

**Comment by adnanmunawar** [comment]
Hello,

For your first question: 

> Is it the value of joint error in the discrete set: {-5., -2., 0 , 2., 5.} in the evaluation?

No, the value is going to be any number between +- 5 degrees for the first two joints, and +-5 mm (or 0.05 simulation units) for the third joint. So for example, we could introduce a joint error of -2.9 deg, 4.1 deg, and 0.5 mm for the first three joints respectively.

For your second point

> Are there joint error for only Joint 1 and 2 of PSM in the evaluation? Code implementation can be found [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/a85dd0271d1b11b6f964d5f46908f2f4060ddcdf/scripts/surgical_robotics_challenge/psm_arm.py#L103). I got little bit confused by the https://github.com/collaborative-robotics/surgical_robotics_challenge/discussions/36#discussioncomment-2511297 since you mentioned that Joint 3, the insertion joint, will also have the joint error.

Thanks for pointing this out. Yes, as quoted in the [discussion](https://github.com/collaborative-robotics/surgical_robotics_challenge/discussions/36#discussioncomment-2511297) that you referenced, the joint errors are going to be added to the first three joints, however, currently they are only being added to the first two. We shall add the third joint error shortly, meanwhile, you can also add it yourself.

Hope that answers your questions.

**Comment by adnanmunawar** [comment, marked as answer]
Hi Hongbin,
We have added the implementation for modeling the joint errors for the first three joints in PR #43. To test and better develop your algorithms, you can dynamically change the joint errors (i.e. directly set their value rather than random generation) by using the ROS topic interface before starting to control the PSMs. The ROS topic is `/ambf/env/<psm_arm_type>/errors_model/set_errors`, where `<psm_arm_type>` is either `psm1` or `psm2`. Check the [implementation](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/scripts/surgical_robotics_challenge/utils/joint_errors_model.py#L56) for more details

**Reply by linhongbin** [reply]
Hi Adnan, thanks for your detailed reply. My questions have been well answered. Looking forward to using the new feature!
